marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/ConstraintsSpec.hs

```diff
@@ -0,0 +1,178 @@
+
+    let
+      assetId :: Chain.AssetId
+      assetId = Chain.AssetId policyId tokenName
+
+      actual :: Constraints.TxConstraints 'Core.Api.V1
+      actual = Constraints.mustSpendRoleToken assetId
+
+      expected :: Constraints.TxConstraints 'Core.Api.V1
+      expected = emptyTxConstraints
+        { roleTokenConstraints = Constraints.SpendRoleTokens $ Set.singleton assetId }
+
+    pure $ actual `shouldBe` expected
+
+  Hspec.describe "adjustTxForMinUtxo" do
+
+    Hspec.QuickCheck.prop "The Marlowe output has NOT been adjusted" do
+      protocol <- hedgehog genProtocolParameters
+
+      marloweScriptHash <- hedgehog genScriptHash
+      let
+        scriptAddress hash = fromCardanoAddressAny
+          $ C.AddressShelley
+          $ C.makeShelleyAddress C.Mainnet (C.PaymentCredentialByScript hash) C.NoStakeAddress
+        marloweAddress = scriptAddress marloweScriptHash
+
+        getValueAtAddress :: Chain.Address -> [C.TxOut C.CtxTx C.BabbageEra] -> Maybe (C.TxOutValue C.BabbageEra)
+        getValueAtAddress targetAddress txos = getFirst . mconcat . map (First
+          . (\(C.TxOut addressInEra txOutValue _ _) ->
+              if fromCardanoAddressInEra C.BabbageEra addressInEra == targetAddress
+                then Just txOutValue
+                else Nothing))
+          $ txos
+
+      txBodyContent <- hedgehog $ genTxBodyContent C.BabbageEra
+
+      let
+        actual = (getValueAtAddress marloweAddress . C.txOuts)
+          <$> Constraints.adjustTxForMinUtxo protocol marloweAddress txBodyContent
+        expected :: Either (Constraints.ConstraintError 'Core.Api.V1) (Maybe (C.TxOutValue C.BabbageEra))
+          = Right $ getValueAtAddress marloweAddress $ C.txOuts txBodyContent
+
+      pure $ actual `shouldBe` expected
+
+    Hspec.QuickCheck.prop "All outputs satisfy Cardano API minimum UTxO requirements" do
+      marloweScriptHash <- hedgehog genScriptHash
+
+      let
+        scriptAddress hash = fromCardanoAddressAny
+          $ C.AddressShelley
+          $ C.makeShelleyAddress C.Mainnet (C.PaymentCredentialByScript hash) C.NoStakeAddress
+        marloweAddress = scriptAddress marloweScriptHash
+
+        valueMeetsMinimumReq :: C.TxOut C.CtxTx C.BabbageEra -> Bool
+        valueMeetsMinimumReq txout@(C.TxOut _ txOrigValue _ _) =
+          case C.calculateMinimumUTxO C.ShelleyBasedEraBabbage txout protocolTestnet of
+            Right minValueFromApi ->
+              (C.selectLovelace . C.txOutValueToValue $ txOrigValue)
+              >= (C.selectLovelace minValueFromApi)
+            Left _ -> False
+
+      txBodyContent <- hedgehog $ genTxBodyContent C.BabbageEra
+
+      pure $ case Constraints.adjustTxForMinUtxo protocolTestnet marloweAddress txBodyContent of
+        Right newTxBodyContent ->
+          all (== True) (map valueMeetsMinimumReq $ C.txOuts newTxBodyContent) `shouldBe` True
+        Left _ -> expectationFailure "Minimum UTxO requirements not met"
```

Comment:
Let's add the error and the actual new outputs to the failure context:
```suggestion
      pure $ case Constraints.adjustTxForMinUtxo protocolTestnet marloweAddress txBodyContent of
        Right newTxBodyContent ->
          counterexample
            (show $ C.txOuts newTxBodyContent)
            (all (== True) (map valueMeetsMinimumReq $ C.txOuts newTxBodyContent) `shouldBe` True)
        Left err -> counterexample (show err) False
```

Reply:
I just made changes to collect up failures in the `valueMeetsMinimumReq` function and be clearer about what exactly failed if there's failure. So, more extensive than the above suggestion, if you think that works?